subsys/bluetooth/host/hci_core.c

```diff
@@ -3133,6 +3133,34 @@ static void read_buffer_size_complete(struct net_buf *buf)
 #endif /* !defined(CONFIG_BT_CLASSIC) */
 #endif /* CONFIG_BT_CONN */
 
+#if defined(CONFIG_BT_CONN) || defined(CONFIG_BT_ISO)
+static void check_num_of_acl_iso_pkts(uint8_t num_acl_pkts, uint8_t num_iso_pkts)
+{
+	/** This check is added to warn if the number of ACL or ISO packets in the controller is not
+	 * equal to the number of bt_conn_tx contexts allocated by the host. The inequality of these
+	 * two values can lead to inefficient resources usage either on the host's or controller's
+	 * side.
+	 */
+#if defined(CONFIG_BT_CONN)
+	CHECKIF(bt_conn_tx_acl_count() != num_acl_pkts) {
+		LOG_WRN("Num of Controller's ACL packets != ACL bt_conn_tx contexts (%u != %u)",
+			num_acl_pkts, bt_conn_tx_acl_count());
+	}
+#else
+	ARG_UNUSED(num_acl_pkts);
+#endif
+
+#if defined(CONFIG_BT_ISO)
+	CHECKIF(bt_conn_tx_iso_count() != num_iso_pkts) {
+		LOG_WRN("Num of Controller's ISO packets != ISO bt_conn_tx contexts (%u != %u)",
+			num_iso_pkts, bt_conn_tx_iso_count());
+	}
```

Comment:
I'm not sure this is a good approach. 

`CONFIG_BT_ISO_TX_BUF_COUNT` refer to the number of SDUs the host supports. `bt_conn_tx_iso_count` refers to the number of HCI ISO data packet the host can enqueue in the controller. Those are not the same.

E.g. if the HCI ISO data buffer size is 100, and the ISO SDUs being sent are 400 octets, then the host will send 4 HCI ISO data buffers to send a single SDU. The number of SDU buffers and HCI ISO data buffers are not directly comparable. 

If we want to add anything here, it should be that 

```suggestion
	CHECKIF(bt_conn_tx_iso_count() > num_iso_pkts) {
		LOG_WRN("Num of Controller's ISO packets < ISO bt_conn_tx contexts (%u < %u)",
			num_iso_pkts, bt_conn_tx_iso_count());
	}
```

As supporting more SDUs than we have HCI ISO data buffers will effectively be wasted memory, except that it would allow an application to enqueue SDUs in the host, while waiting for HCI buffers to be free'd.

Reply:
If `CONFIG_BT_ISO_TX_BUF_COUNT` and `bt_conn_tx_iso_count` aren't the same, why Host needs to allocate `CONFIG_BT_ISO_TX_BUF_COUNT` amount of `bt_conn_tx` objects? https://github.com/zephyrproject-rtos/zephyr/blob/e96bf44d934050b404e53aefdcb9756290b4c18e/subsys/bluetooth/host/conn.c#L196

From what I see, number of `bt_conn_tx` objects for ISO (which is `iso_tx` array) should be equal to number of ISO data packets Controller can enqueue (`CONFIG_BT_CTLR_ISO_TX_BUFFERS`) (not more, not less), otherwise it will be a waste of resources for either Host or Controller. Host picks free `bt_conn_tx` object only if it can send more data to Controller:
https://github.com/zephyrproject-rtos/zephyr/blob/e96bf44d934050b404e53aefdcb9756290b4c18e/subsys/bluetooth/host/conn.c#L944-L956

Reply:
It sounds like `iso_tx` should have own Kconfig then.

Reply:
I think there may be a larger issue here w.r.t. fragmentation of ISO SDUs. 

If we keep the example an ISO SDU of size 400, where the HCI ISO data buffer size from the controller is 100. 
Then for each SDU (`CONFIG_BT_ISO_TX_BUF_COUNT` denotes the number of SDUs supported), we need 4 HCI ISO data buffers to send (since the SDU size is completely dynamic, this is not something we can add a simple one-time check for). 
There should only be 1 `bt_conn_tx` context assigned for this, as that basically is what handles the callback to the application 1 SDU => 1 call to the callback. 

However `tx_notify_process` seems to assume that 1 num_complete_packets event means that the SDU has been sent, and the `bt_conn_tx` can be free'd and the callback can be called. That may be the case for ACL (not sure), but certainly not the case for ISO where SDUs can be fragmented over HCI. So a single SDU may require 4 (or any number really) HCI ISO data packets, and will get 4 HCI num_complete_packets, but should only require 1 `bt_conn_tx` and only call the application provided callback once. 

So in my understanding:
Number of SDUs supported == CONFIG_BT_ISO_TX_BUF_COUNT != HCI ISO data buffer count

If the HCI ISO data buffer count is smaller than the number of SDUs in the host, then that's OK as the SDUs are just enqueue in the host (but could possibly be a waste of memory)
If the HCI ISO data buffer count is larger than then number of SDUs in the host, then that's OK, as a single SDU may use one or more HCI ISO data buffers

Reply:
It seems to be a misunderstanding here.

After https://github.com/zephyrproject-rtos/zephyr/pull/72090 `bt_conn_tx` is always allocated when Host sends anything to Controller regardless whether the callback is initialized or set to NULL.

If fragmentation is done in conn.c when `ISO_Data_Packet_Length` returned by Controller less than size of packet Host wants to send, then the callback will be called only for the last fragment:
https://github.com/zephyrproject-rtos/zephyr/blob/8b1f06af052f465e7d7094a49764d6aba76b4655/subsys/bluetooth/host/conn.c#L1094-L1102

If, however, the packet is not fragmented by `ISO_Data_Packet_Length`, then the callback will be called (if set) upon Number Of Completed Packets event received from Controller.

So, if ISO supplied a callback for the buffer when host calls `conn->get_and_clear_cb()`, the callback will be called. And I see that ISO currently always supplies a callback: https://github.com/zephyrproject-rtos/zephyr/blob/8b1f06af052f465e7d7094a49764d6aba76b4655/subsys/bluetooth/host/iso.c#L162-L166

For L2CAP, for example, this is done for the last fragment for dynamic channels (when L2CAP fragments SDUs into PDUs): https://github.com/zephyrproject-rtos/zephyr/blob/8b1f06af052f465e7d7094a49764d6aba76b4655/subsys/bluetooth/host/l2cap.c#L983-L991
and always for static ones: https://github.com/zephyrproject-rtos/zephyr/blob/8b1f06af052f465e7d7094a49764d6aba76b4655/subsys/bluetooth/host/l2cap.c#L756

So, if you say there should be only 1 `bt_conn_tx` object for ISO, then it is ISO responsibility to do so after https://github.com/zephyrproject-rtos/zephyr/pull/72090.

Reply:
Thanks for the clarification. At least that ensures that 1 SDU => 1 Callback. 

But it still seems that the current code expects that 1 HCI ISO data packet => 1 bt_conn_tx. 
If that is the case, then we cannot do 
```c
static struct bt_conn_tx iso_tx[CONFIG_BT_ISO_TX_BUF_COUNT];
```
if `CONFIG_BT_ISO_TX_BUF_COUNT` represents SDUs and not HCI ISO data packets.

The definition of `BT_ISO_TX_BUF_COUNT` is

```c
config BT_ISO_TX_BUF_COUNT
	int "Number of Isochronous TX buffers"
	default 1
	range 1 $(UINT8_MAX)
	help
	  Number of buffers available for outgoing Isochronous channel SDUs.
```

Which does not explicitly state that it's number of SDUs, but rather number of buffers _for_ SDUs. Most tests/samples use it as the SDU count, but they also all assume that no fragmentation is done. The only place where it is actually used as a SDU count is in hci_raw.c: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_raw.c#L58-L61, but it also seems that hci_raw treats `CONFIG_BT_BUF_ACL_TX_COUNT` as ACL SDU count (given the size of the pool items (which are larger than HCI data packets). 

We haven't had much code coverage (being added in https://github.com/zephyrproject-rtos/zephyr/pull/81871), so it kind of makes sense that this (wrong) assumption that 1 SDU == 1 HCI ISO data packet has gone unnoticed for a long time. 

That being said, and have been refreshed on the buffer stuff, I think the changes here are OK for ISO. Thanks for your patience :)